core/src/main/java/org/apache/iceberg/rest/requests/PlanTableScanRequestParser.java

```diff
@@ -0,0 +1,142 @@
+  private static final String FILTER = "filter";
+  private static final String CASE_SENSITIVE = "case-sensitive";
+  private static final String USE_SNAPSHOT_SCHEMA = "use-snapshot-schema";
+  private static final String START_SNAPSHOT_ID = "start-snapshot-id";
+  private static final String END_SNAPSHOT_ID = "end-snapshot-id";
+  private static final String STATS_FIELDS = "stats-fields";
+
+  private PlanTableScanRequestParser() {}
+
+  public static String toJson(PlanTableScanRequest request) {
+    return toJson(request, false);
+  }
+
+  public static String toJson(PlanTableScanRequest request, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(request, gen), pretty);
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static void toJson(PlanTableScanRequest request, JsonGenerator gen) throws IOException {
+    Preconditions.checkArgument(null != request, "Invalid request: planTableScanRequest null");
+
+    if (request.snapshotId() != null
+        || request.startSnapshotId() != null
+        || request.endSnapshotId() != null) {
+      Preconditions.checkArgument(
+          request.snapshotId() != null
+              ^ (request.startSnapshotId() != null && request.endSnapshotId() != null),
+          "Either snapshotId must be provided or both startSnapshotId and endSnapshotId must be provided");
+    }
+
+    gen.writeStartObject();
+    if (request.snapshotId() != null) {
+      gen.writeNumberField(SNAPSHOT_ID, request.snapshotId());
+    }
+
+    if (request.startSnapshotId() != null) {
+      gen.writeNumberField(START_SNAPSHOT_ID, request.startSnapshotId());
+    }
+
+    if (request.endSnapshotId() != null) {
+      gen.writeNumberField(END_SNAPSHOT_ID, request.endSnapshotId());
+    }
+
+    if (request.select() != null && !request.select().isEmpty()) {
+      JsonUtil.writeStringArray(SELECT, request.select(), gen);
+    }
+
+    if (request.filter() != null) {
+      gen.writeStringField(FILTER, ExpressionParser.toJson(request.filter()));
+    }
+
+    gen.writeBooleanField(CASE_SENSITIVE, request.caseSensitive());
+    gen.writeBooleanField(USE_SNAPSHOT_SCHEMA, request.useSnapshotSchema());
+
+    if (request.statsFields() != null && !request.statsFields().isEmpty()) {
+      JsonUtil.writeStringArray(STATS_FIELDS, request.statsFields(), gen);
+    }
+
+    gen.writeEndObject();
+  }
+
+  public static PlanTableScanRequest fromJson(String json) {
+    return JsonUtil.parse(json, PlanTableScanRequestParser::fromJson);
+  }
+
+  public static PlanTableScanRequest fromJson(JsonNode json) {
+    Preconditions.checkArgument(null != json, "Invalid request: planTableScanRequest null");
+
+    Long snapshotId = JsonUtil.getLongOrNull(SNAPSHOT_ID, json);
+    Long startSnapshotId = JsonUtil.getLongOrNull(START_SNAPSHOT_ID, json);
+    Long endSnapshotId = JsonUtil.getLongOrNull(END_SNAPSHOT_ID, json);
+
+    if (snapshotId != null || startSnapshotId != null || endSnapshotId != null) {
```

Comment:
this can be removed once the validation logic is moved into `validate()`

Reply:
will fix